### PR TITLE
Test automation for User Story 2

### PR DESCRIPTION
**Pull request by poornimadayalan**
User Story 2:
---------------
Description:
--------------
Automate the test cases for language and skills. Remember to use the Specflow framework with BDD (Given,
When, Then etc with step definitions and pages please refer to Specflow docs mentioned above for more info),
you may use the following http://git.mvp.studio/qa-examples/onboarding.specflow provided. Note: please change the
credentials in the solution file according to your details.

Use case:
-----------
As a user I would be able to show what languages and skills I know.
So that the people seeking for skills and languages can look at what details I hold.

Acceptance Criteria:
-----------------------
o Write automation tests from the test cases that were written for language and skills.
o Make sure all tests pass.

### CODE REVIEW

File: Features/MyProfileLanguages.feature [hunk 1]
```diff
@@ -0,0 +1,38 @@
+Feature: This test suite runs my profile language related test scenarios
+
+Scenario: 1. Verify user is able to add known languages with proficiency level
+	Given User logs into Mars application
+	When User adds 'English' language with 'Fluent' proficiency level into the known languages list
+	Then The 'English' language with 'Fluent' proficiency level should appear in the known languages list
+
+Scenario: 2. Verify user is able to edit an existing known language with proficiency level
```
Comment (amanbansalnz) [inline Features/MyProfileLanguages.feature]: Please use tables  and pass values to your scenarios for example look at the below scenario 
This way you will get exposure to use them  and will know how to pass test data to the tests


Example 

	Given I logged into the portal
	When I click on tab Education
	And I add my education 
	Then I am able to see my education'

Would  change to 


@Education
Scenario Outline: 1-Add education
	Given I logged into the portal
	When I click on tab Education
	And I add my education including '<Country>', '<University>', '<Title>', '<Degree>', '<Graduation Year>'
	Then I am able to see my education details including '<Country>', '<University>', '<Title>', '<Degree>', '<Graduation Year>'

	Examples:
	| Country     | University                      | Title  | Degree  | Graduation Year |
	| New Zealand | Whitireia Community Polytechnic | M.Tech | Diploma | 2021            |
	| Venezuela   | Hhutech University              | B.A    | Diploma | 2011            |

See  https://cucumber.io/docs/gherkin/reference/

File: Features/MyProfileLanguages.feature [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	Given User logs into Mars application
+	When User edits the 'English' language with 'Basic' proficiency level from the known languages list
+	Then The proficiency level for 'English' language should be updated to 'Basic'
+
```
Comment (amanbansalnz) [inline Features/MyProfileLanguages.feature:17]: Please increase your coverage around the features you  are testing here is an idea of what is missing 

• Basic positive tests (happy paths)
• Extended positive testing with optional parameters
• Negative testing with valid input
• Negative testing with invalid input
• Destructive testing
• Security, authorisation, and permission tests (which are out of the scope of this post)

E.g of above scenarios

Happy path tests check basic functionality and the acceptance criteria of the API. We later extend positive tests to include optional parameters and extra functionality.

The next group of tests is negative testing where we expect the application to gracefully handle problem scenarios with both valid user input (for example, trying to add an existing username) and invalid user input (trying to add a username which is null).

Destructive testing is a deeper form of negative testing where we intentionally attempt to break the API to check its robustness (for example, sending a huge payload body in an attempt to overflow the system).

If some scenarios does not make sense to add you can you should know why it does not make  sense to add.

File: Pages/ProfilePageLanguages.cs [hunk 3]
```diff
@@ -0,0 +1,239 @@
+﻿using NLog;
+using OpenQA.Selenium;
+using ProjectMars.Constants;
+using ProjectMars.Utilities.Hooks;
+
+namespace ProjectMars.Pages
+{
+    /// <summary>
+    /// Class for profile page language related actions.
+    /// </summary>
```
Comment (amanbansalnz) [inline Pages/ProfilePageLanguages.cs:9]: so i see your web elements are in your constant file, that is not correct! Ideally it makes sense that you keep the webelements with the pages this way you see what elements are present on what page! Please change this and also make sure you create a skills and langauge page

File: StepDefinitions/MyProfileLanguagesStep.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+namespace ProjectMars.StepDefinitions
+{
+    /// <summary>
+    /// Step definition for language functionality in my profile.
+    /// </summary>
+    [Binding]
+    public class MyProfileLanguagesStep : Hooks
+    {
+        SignInPage signInPageObj = new SignInPage();
```
Comment (amanbansalnz) [inline StepDefinitions/MyProfileLanguagesStep.cs:14]: please make use of constructor Please see 
https://www.knowledgehut.com/tutorials/csharp/csharp-constructors

File: StepDefinitions/MyProfileSkillsStep.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+using TechTalk.SpecFlow;
+
+namespace ProjectMars.StepDefinitions
+{
+    // <summary>
+    /// Step definition for skill functionality in my profile.
+    /// </summary>
+    public class MyProfileSkillsStep : Hooks
+    {
+        ProfilePageSkills profilePageSkillObj = new ProfilePageSkills();
```
Comment (amanbansalnz) [inline StepDefinitions/MyProfileSkillsStep.cs:14]: Please see 
https://www.knowledgehut.com/tutorials/csharp/csharp-constructors

File: StepDefinitions/MyProfileLanguagesStep.cs [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        [When(@"User adds '([^']*)' language with '([^']*)' proficiency level into the known languages list")]
+        public void WhenUserAddsLanguageWithProficiencyLevelIntoTheKnownLanguagesList(string language, string proficiencyLevel)
+        {
+            profilePageObj.AddLanguageWithProficiencyLevel(driver, language, proficiencyLevel);
+        }
+
+        [Then(@"The '([^']*)' language with '([^']*)' proficiency level should appear in the known languages list")]
+        public void ThenTheLanguageWithProficiencyLevelShouldAppearInTheKnownLanguagesList(string language, string proficiencyLevel)
+        {
+            if (profilePageObj.VerifyTheAddedLanguage(driver, language, proficiencyLevel))
```
Comment (amanbansalnz) [inline StepDefinitions/MyProfileLanguagesStep.cs]: i would also assert the message that you get when you add skill or language

Comment (amanbansalnz) [inline StepDefinitions/MyProfileLanguagesStep.cs]: please do not pass the driver in the action methodprofilePageObj.VerifyTheAddedLanguage(driver, language, proficiencyLevel)

Comment (amanbansalnz) [reply StepDefinitions/MyProfileLanguagesStep.cs]: please fix this for all action methods

File: Pages/ProfilePageLanguages.cs [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    public class ProfilePageLanguages : Hooks
```
Comment (amanbansalnz) [inline Pages/ProfilePageLanguages.cs]: you could extend the class to Driver rather than Hooks

File: StepDefinitions/MyProfileLanguagesStep.cs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        ProfilePageLanguages profilePageObj = new ProfilePageLanguages();
+
+        [Given(@"User logs into Mars application")]
+        public void GivenUserLogsIntoMarsApplication()
+        {
+            signInPageObj.SignInToApplication(driver);
+        }
+
+        [When(@"User adds '([^']*)' language with '([^']*)' proficiency level into the known languages list")]
+        public void WhenUserAddsLanguageWithProficiencyLevelIntoTheKnownLanguagesList(string language, string proficiencyLevel)
```
Comment (amanbansalnz) [inline StepDefinitions/MyProfileLanguagesStep.cs:25]: you need to think about state 


example before adding language you need to make sure there a no existing languages as this could result in false positive test pass
you will need to think about this for all the scenarios

the state is say that you have a test that is adding a language english, the second time you run this test you should delete any existing language so that the state is reset, this will guarantee that your test does not produce false negatives and passes always
you can assume that test execution order is not always guaranteed
imagine if you run the tests  in parallel mode the all tests will run at the same time
so you need to also make sure you do not wipe out other tests data
hence why when you remove data the best thing to do is delete only data you added for the test to be correct
e.g test 1 added english and french as language test 2 added hindi then when you reset data you already know what language you are adding for each test
there for on reset of the two test you should delete english and french for test 1 and hindi for test 2

So to sum up you should delete all data if there is any before the start of any test and then after each test you should delete the data that a particular test is adding!
This is ensure that upon staring the test pack you have reset state and upon finishing each test deletes their data that they added!

**Review by amanbansalnz — CHANGES_REQUESTED** [review]
please make the suggested changes

File: Pages/ProfilePageLanguages.cs (hunk 7 above)
Comment (poornimadayalan) [reply Pages/ProfilePageLanguages.cs]: Thanks, Aman, for all your feedback! I'll make sure to address the comments you mentioned and update the changes. Once done, I'll submit a new pull request. Thanks again!

File: StepDefinitions/MyProfileLanguagesStep.cs (hunk 4 above)
Comment (amanbansalnz) [reply StepDefinitions/MyProfileLanguagesStep.cs:14]: you have not done this?

File: StepDefinitions/MyProfileLanguagesStep.cs (hunk 8 above)
Comment (amanbansalnz) [reply StepDefinitions/MyProfileLanguagesStep.cs:25]: you have not done this?

**Review by amanbansalnz — CHANGES_REQUESTED** [review]
Hey Poornima so you have not addressed a couple of review comments can you please address them the rest looks good

File: StepDefinitions/MyProfileLanguagesStep.cs (hunk 8 above)
Comment (poornimadayalan) [reply StepDefinitions/MyProfileLanguagesStep.cs:25]: Screenshot 2024-05-23 083956
Screenshot 2024-05-23 084040
Thanks for reviewing my code. Yes, I've made the changes you suggested. I cleared all the data before moving to the next scenario and also did a cleanup test before the test started.

File: StepDefinitions/MyProfileLanguagesStep.cs (hunk 4 above)
Comment (poornimadayalan) [reply StepDefinitions/MyProfileLanguagesStep.cs:14]: I used the constructor as well. As per my understanding, I have implemented it . I've attached a screenshot for your reference. If it's not correct, please let me know how to fix it. Thanks for your help!

Screenshot 2024-05-23 090217
Screenshot 2024-05-23 090236

File: StepDefinitions/MyProfileLanguagesStep.cs (hunk 6 above)
Comment (poornimadayalan) [reply StepDefinitions/MyProfileLanguagesStep.cs]: Yes, I have asserted all the messages.

Screenshot 2024-05-23 092455
